wasm/prover/src/prover.rs

```diff
@@ -0,0 +1,482 @@
+use futures::channel::oneshot;
+use std::ops::Range;
+use std::panic;
+use tlsn_prover::tls::{Prover, ProverConfig};
+use wasm_bindgen_futures::spawn_local;
+use web_time::Instant;
+
+use ws_stream_wasm::*;
+
+use crate::request_opt::RequestOptions;
+use crate::requests::{ClientType, NotarizationSessionRequest, NotarizationSessionResponse};
+
+pub use wasm_bindgen_rayon::init_thread_pool;
+
+pub use crate::request_opt::VerifyResult;
+use crate::{fetch_as_json_string, log};
+use futures::AsyncWriteExt;
+use hyper::{body::to_bytes, Body, Request, StatusCode};
+use js_sys::Array;
+use strum::EnumMessage;
+use tlsn_core::proof::TlsProof;
+use tokio_util::compat::FuturesAsyncReadCompatExt;
+use url::Url;
+use wasm_bindgen::prelude::*;
+use web_sys::{Headers, RequestInit, RequestMode};
+
+#[derive(strum_macros::EnumMessage, Debug, Clone, Copy)]
+#[allow(dead_code)]
+enum ProverPhases {
+    #[strum(message = "Connect application server with websocket proxy")]
+    ConnectWsProxy,
+    #[strum(message = "Build prover config")]
+    BuildProverConfig,
+    #[strum(message = "Set up prover")]
+    SetUpProver,
+    #[strum(message = "Bind the prover to the server connection")]
+    BindProverToConnection,
+    #[strum(message = "Spawn the prover thread")]
+    SpawnProverThread,
+    #[strum(message = "Attach the hyper HTTP client to the TLS connection")]
+    AttachHttpClient,
+    #[strum(message = "Spawn the HTTP task to be run concurrently")]
+    SpawnHttpTask,
+    #[strum(message = "Build request")]
+    BuildRequest,
+    #[strum(message = "Start MPC-TLS connection with the server")]
+    StartMpcConnection,
+    #[strum(message = "Received response from the server")]
+    ReceivedResponse,
+    #[strum(message = "Parsing response from the server")]
+    ParseResponse,
+    #[strum(message = "Close the connection to the server")]
+    CloseConnection,
+    #[strum(message = "Start notarization")]
+    StartNotarization,
+    #[strum(message = "Commit to data")]
+    Commit,
+    #[strum(message = "Finalize")]
+    Finalize,
+    #[strum(message = "Notarization complete")]
+    NotarizationComplete,
+    #[strum(message = "Create Proof")]
+    CreateProof,
+}
+
+fn log_phase(phase: ProverPhases) {
+    log!(
+        "!@# tlsn-js {}: {}",
+        phase as u8,
+        phase.get_message().unwrap()
+    );
+}
+
+#[wasm_bindgen]
+pub async fn prover(
```

Comment:
At some point we should invest the effort and break this down into smaller functions.

Reply:
Indeed. But waiting on @0xtsukino's Extism experiments first